crates/solver/src/liquidity/zeroex.rs

```diff
@@ -9,39 +9,80 @@ use {
         liquidity_collector::LiquidityCollecting,
         settlement::SettlementEncoder,
     },
-    anyhow::Result,
+    anyhow::{Context, Result},
     contracts::{GPv2Settlement, IZeroEx},
+    ethcontract::Address,
+    futures::{SinkExt, StreamExt},
     model::{order::OrderKind, TokenPair},
     primitive_types::{H160, U256},
+    reqwest::Url,
     shared::{
         ethrpc::Web3,
         http_solver::model::TokenAmount,
         recent_block_cache::Block,
-        zeroex_api::{Order, OrderRecord, OrdersQuery, ZeroExApi},
+        zeroex_api::{
+            websocket::{OrderRecord, OrdersResponse},
+            Order,
+            OrdersQuery,
+            ZeroExApi,
+        },
     },
     std::{
         collections::{HashMap, HashSet},
         sync::Arc,
+        time::Duration,
     },
+    tokio::{net::TcpStream, sync::Mutex},
+    tokio_tungstenite::{MaybeTlsStream, WebSocketStream},
 };
 
 pub struct ZeroExLiquidity {
-    pub api: Arc<dyn ZeroExApi>,
     pub zeroex: IZeroEx,
     pub gpv2: GPv2Settlement,
     pub allowance_manager: Box<dyn AllowanceManaging>,
+    // cached orders, updated by background task
+    // key for hashmap is order hash
+    pub cache: Arc<Mutex<HashMap<String, OrderRecord>>>,
```

Comment:
I've added type alias, since 0x returns id as some hash that we use as a String anyway in the rest of the code so I don't know how differently to define it's type.